Firestore/core/src/firebase/firestore/core/query_listener.mm

```diff
@@ -0,0 +1,146 @@
+                            snapshot.from_cache(),
+                            snapshot.sync_state_changed(),
+                            /*excludes_metadata_changes=*/true};
+  }
+
+  if (!raised_initial_event_) {
+    if (ShouldRaiseInitialEvent(snapshot, online_state_)) {
+      RaiseInitialEvent(snapshot);
+    }
+  } else if (ShouldRaiseEvent(snapshot)) {
+    listener_(snapshot);
+  }
+
+  snapshot_ = std::move(snapshot);
+}
+
+void QueryListener::OnError(Status error) {
+  listener_(std::move(error));
+}
+
+void QueryListener::OnOnlineStateChanged(OnlineState online_state) {
+  online_state_ = online_state;
+  if (snapshot_.has_value() && !raised_initial_event_ &&
+      ShouldRaiseInitialEvent(snapshot_.value(), online_state)) {
+    RaiseInitialEvent(snapshot_.value());
+  }
+}
+
+bool QueryListener::ShouldRaiseInitialEvent(const ViewSnapshot& snapshot,
+                                            OnlineState online_state) const {
+  HARD_ASSERT(!raised_initial_event_, "Determining whether to raise initial "
+                                      "event, but already had first event.");
+
+  // Always raise the first event when we're synced
+  if (!snapshot.from_cache()) {
+    return true;
+  }
+
+  // NOTE: We consider OnlineState.Unknown as online (it should become Offline
+  // or Online if we wait long enough).
+  bool maybe_online = online_state != OnlineState::Offline;
+
+  // Don't raise the event if we're online, aren't synced yet (checked
+  // above) and are waiting for a sync.
+  if (options_.wait_for_sync_when_online() && maybe_online) {
+    HARD_ASSERT(snapshot.from_cache(),
+                "Waiting for sync, but snapshot is not from cache.");
+    return false;
+  }
+
+  // Raise data from cache if we have any documents or we are offline
+  return !snapshot.documents().empty() || online_state == OnlineState::Offline;
+}
+
+bool QueryListener::ShouldRaiseEvent(const ViewSnapshot& snapshot) const {
+  // We don't need to handle includeDocumentMetadataChanges here because the
+  // Metadata only changes have already been stripped out if needed. At this
+  // point the only changes we will see are the ones we should propagate.
+  if (!snapshot.document_changes().empty()) {
+    return true;
+  }
+
+  bool has_pending_writes_changed =
+      snapshot_.has_value() &&
+      snapshot_.value().has_pending_writes() != snapshot.has_pending_writes();
+  if (snapshot.sync_state_changed() || has_pending_writes_changed) {
+    return options_.include_query_metadata_changes();
+  }
+
+  // Generally we should have hit one of the cases above, but it's possible to
+  // get here if there were only metadata docChanges and they got stripped out.
+  return false;
+}
+
+void QueryListener::RaiseInitialEvent(const ViewSnapshot& snapshot) {
+  HARD_ASSERT(!raised_initial_event_,
+              "Trying to raise initial events for second time");
+  ViewSnapshot modified_snapshot = ViewSnapshot::FromInitialDocuments(
```

Comment:
Very optional: add a newline above this line to separate preconditions checking from the main logic?

Reply:
Done.